src/AutoMapper/Mappers/ReadOnlyDictionaryMapper.cs

```diff
@@ -0,0 +1,46 @@
+﻿using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Linq.Expressions;
+using AutoMapper.Configuration;
+using AutoMapper.Internal;
+using AutoMapper.Mappers.Internal;
+
+namespace AutoMapper.Mappers
+{
+    using static Expression;
+    using static ExpressionFactory;
+    using static CollectionMapperExpressionFactory;
+
+    public class ReadOnlyDictionaryMapper : IObjectMapper
+    {
+        public bool IsMatch(TypePair context)
+        {
+            if (!(context.SourceType.IsEnumerableType() && context.DestinationType.IsGenericType()))
+                return false;
+
+            var genericType = context.DestinationType.GetGenericTypeDefinition();
+
+            return genericType == typeof(ReadOnlyDictionary<,>) || genericType == typeof(IReadOnlyDictionary<,>);
+        }
+
+        public Expression MapExpression(IConfigurationProvider configurationProvider, ProfileMap profileMap,
+            IMemberMap memberMap, Expression sourceExpression, Expression destExpression, Expression contextExpression)
+        {
+            var dictionaryTypes = ElementTypeHelper.GetElementTypes(destExpression.Type, ElementTypeFlags.BreakKeyValuePair);
+            var dictType = typeof(Dictionary<,>).MakeGenericType(dictionaryTypes);
+            var dict = MapCollectionExpression(configurationProvider, profileMap, memberMap, sourceExpression, Default(dictType), contextExpression, typeof(Dictionary<,>), MapKeyPairValueExpr);
+            var dest = Variable(dictType, "dest");
+
+            var readOnlyDictType = destExpression.Type.IsInterface
+                ? typeof(ReadOnlyDictionary<,>).MakeGenericType(dictionaryTypes)
+                : destExpression.Type;
+
+            return Block(new[] { dest }, 
+                Assign(dest, dict), 
+                Condition(NotEqual(dest, Default(dictType)), 
+                    ToType(New(readOnlyDictType.GetDeclaredConstructors().First(), dest), destExpression.Type), 
```

Comment:
Maybe a more precise constructor filter? The same in CreateReadOnlyCollection.

Reply:
Done. Went with more precise "look for a ctor that matches exactly"